src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/CompletionItemSource.cs

```diff
@@ -0,0 +1,249 @@
+            {
+                return EditorCompletion.CompletionStartData.DoesNotParticipateInCompletion;
+            }
+
+            var service = document.GetLanguageService<CompletionService>();
+            if (service == null)
+            {
+                return EditorCompletion.CompletionStartData.DoesNotParticipateInCompletion;
+            }
+
+            var sourceText = document.GetTextSynchronously(cancellationToken);
+
+            // TODO: Check CompletionOptions.TriggerOnTyping  https://github.com/dotnet/roslyn/issues/27427
+            if (!(trigger.Reason == EditorCompletion.CompletionTriggerReason.Invoke ||
+                trigger.Reason == EditorCompletion.CompletionTriggerReason.InvokeAndCommitIfUnique)
+                && !service.ShouldTriggerCompletion(sourceText, triggerLocation.Position, GetRoslynTrigger(trigger, triggerLocation)))
+            {
+                return EditorCompletion.CompletionStartData.DoesNotParticipateInCompletion;
+            }
+
+            return new EditorCompletion.CompletionStartData(
+                participation: EditorCompletion.CompletionParticipation.ProvidesItems,
+                applicableToSpan: new SnapshotSpan(triggerLocation.Snapshot, service.GetDefaultCompletionListSpan(sourceText, triggerLocation.Position).ToSpan()));
+        }
+
+        public async Task<EditorCompletion.CompletionContext> GetCompletionContextAsync(
+            IAsyncCompletionSession session, 
+            EditorCompletion.CompletionTrigger trigger, 
+            SnapshotPoint triggerLocation, 
+            SnapshotSpan applicableToSpan, 
+            CancellationToken cancellationToken)
+        {
+            var document = triggerLocation.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return new EditorCompletion.CompletionContext(ImmutableArray<EditorCompletion.CompletionItem>.Empty);
+            }
+
+            var completionService = document.GetLanguageService<CompletionService>();
+
+            var completionList = await completionService.GetCompletionsAsync(
+                document,
+                triggerLocation,
+                GetRoslynTrigger(trigger, triggerLocation)).ConfigureAwait(false);
+
+            if (completionList == null)
+            {
+                return new EditorCompletion.CompletionContext(ImmutableArray<EditorCompletion.CompletionItem>.Empty);
+            }
+
+            var filterCache = new Dictionary<string, EditorCompletion.CompletionFilter>();
+
+            var itemsBuilder = ImmutableArray.CreateBuilder<EditorCompletion.CompletionItem>();
+            foreach (var roslynItem in completionList.Items)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                var item = Convert(document, roslynItem, completionService, filterCache);
+                item.Properties.AddProperty(TriggerBuffer, triggerLocation.Snapshot.TextBuffer);
```

Comment:
it worries me that things are holding onto a TextBuffer (a mutable resource) what's the reason for that.  this code should have documentation explaining things.